feat: Add Single Validator RPC Method for Historical Data Retrieval

StoneMac65 opened the pull request:
# Summary

This PR adds a new RPC method `getValidatorStats` to retrieve historical data and performance for a single validator instead of all validators, addressing the need for more efficient validator monitoring as discussed on Discord. Prior to this implementation, retrieving validator performance data required calling `getValidatorsStats`, which returns data for ALL validators.

This PR introduces a targeted approach that:

- Reduces data transfer and processing overhead by 99%+
- Enables focused validator monitoring with ~2-5KB responses
- Supports historical data analysis with optional time range filtering
- Prepares the foundation for future archival functionality

## Core Implementation

1. **New Types & Schemas** (`yarn-project/stdlib/src/validators/`)
   - Added `SingleValidatorStats` type for single validator response
   - Added corresponding Zod schema for validation

2. **RPC Interface** (`yarn-project/stdlib/src/interfaces/aztec-node.ts`)
   - Added `getValidatorStats` method to `AztecNode` interface
   - Added schema definition with optional slot range parameters

3. **Sentinel Implementation** (`yarn-project/aztec-node/src/sentinel/`)
   - Made `SentinelStore.getHistory()` public for single validator access
   - Added `computeStatsForSingleValidator()` method to `Sentinel` class
   - Implemented slot range filtering for historical data analysis

4. **Node Service** (`yarn-project/aztec-node/src/aztec-node/server.ts`)
   - Added RPC endpoint implementation that delegates to Sentinel

## API Details

**Method**: `getValidatorStats`
**Namespace**: `node_getValidatorStats` (in sandbox)

**Parameters**:
- `validatorAddress` (string): The validator's Ethereum address
- `fromSlot` (optional bigint): Start slot for filtering historical data
- `toSlot` (optional bigint): End slot for filtering historical data

**Returns**: `SingleValidatorStats | undefined`

**Response Structure**:
```typescript
{
  validator: ValidatorStats,           // Current validator statistics
  provenPerformance: Array<{           // Historical proven performance
    missed: number,
    total: number,
    epoch: bigint
  }>,
  lastProcessedSlot?: bigint,          // Last processed slot
  initialSlot?: bigint,                // Initial tracking slot
  slotWindow: number                   // Size of tracking window
}
```

## Usage Examples

### Get all data for a validator

```bash
curl -X POST http://localhost:8080 \
  -H "Content-Type: application/json" \
  -d '{"jsonrpc":"2.0","method":"node_getValidatorStats","params":["0x1234..."],"id":1}'
```

### Get data for specific slot range

```bash
curl -X POST http://localhost:8080 \
  -H "Content-Type: application/json" \
  -d '{"jsonrpc":"2.0","method":"node_getValidatorStats","params":["0x1234...", "100", "200"],"id":1}'
```

## Testing

### Development Testing

- **Tag v1.2.0**: Tested and validated against Aztec v1.2.0 tag
- **Current testnet**: Successfully tested on live Aztec testnet
- **Integration tests**: Validated with real validator data and various scenarios

### Docker Image Available

A testnet-compatible version is available for testing:
```bash
docker pull stonemac65/sentinel:v1.2.0
```

### Real Test Server Example on Current Testnet

**Request:**
```bash
curl -X POST http://localhost:8080 \
  -H "Content-Type: application/json" \
  -d '{"jsonrpc": "2.0", "method": "node_getValidatorStats", "params": ["0x2ccf122e8953fd81b547ccb4f27d6dfa13907b03", "24300", "24400"], "id": 1}'
```

**Result for real validator:**
```json
{
  "jsonrpc":"2.0",
  "id":1,
  "result":{
    "validator":{
      "address":"0x2ccf122e8953fd81b547ccb4f27d6dfa13907b03",
      "lastAttestation":{
        "timestamp":"1753965252",
        "slot":"24400",
        "date":"2025-07-31T12:34:12.000Z"
      },
      "totalSlots":14,
      "missedProposals":{"currentStreak":0,"count":0},
      "missedAttestations":{"currentStreak":0,"rate":0,"count":0},
      "history":[
        {"slot":"24385","status":"attestation-sent"},
        {"slot":"24386","status":"attestation-sent"},
        {"slot":"24387","status":"attestation-sent"},
        {"slot":"24388","status":"attestation-sent"},
        {"slot":"24389","status":"attestation-sent"},
        {"slot":"24390","status":"attestation-sent"},
        {"slot":"24391","status":"attestation-sent"},
        {"slot":"24392","status":"attestation-sent"},
        {"slot":"24393","status":"attestation-sent"},
        {"slot":"24394","status":"attestation-sent"},
        {"slot":"24395","status":"attestation-sent"},
        {"slot":"24397","status":"attestation-sent"},
        {"slot":"24399","status":"attestation-sent"},
        {"slot":"24400","status":"attestation-sent"}
      ]
    },
    "provenPerformance":[
      {"epoch":"762","missed":0,"total":23},
      {"epoch":"792","missed":0,"total":14},
      {"epoch":"817","missed":0,"total":14},
      {"epoch":"870","missed":0,"total":20},
      {"epoch":"896","missed":0,"total":21},
      {"epoch":"920","missed":0,"total":25},
      {"epoch":"965","missed":0,"total":21},
      {"epoch":"990","missed":0,"total":24}
    ],
    "lastProcessedSlot":"32040",
    "initialSlot":"24383",
    "slotWindow":960000
  }
}
```

StoneMac65 committed:
Implement single validator stats computation in Sentinel

StoneMac65 committed:
Make SentinelStore getHistory method public

StoneMac65 committed:
Add getValidatorStats RPC method to AztecNode server

StoneMac65 committed:
Add tests for getValidatorStats RPC method

StoneMac65 committed:
Define SingleValidatorStats type

StoneMac65 committed:
Add getValidatorStats method to AztecNode interface

StoneMac65 committed:
Add SingleValidatorStats schema validation

0x4r45h committed:
Add tests for getValidatorStats

StoneMac65 committed:
fix: correct validator address type in getValidatorStats tests

- Change validatorAddress parameter from EthAddress to string in MockAztecNode
- Add .toString() calls when passing EthAddress to getValidatorStats method
- Update expect statements to check for string type instead of EthAddress instance
- Fixes TypeScript compilation errors in test suite

Maddiaa0 commented on yarn-project/stdlib/src/interfaces/aztec-node.ts:
```diff
@@ -577,6 +580,11 @@ export const AztecNodeApiSchema: ApiSchemaFor<AztecNode> = {
 
+  getValidatorStats: z
+    .function()
+    .args(z.string(), optional(schemas.BigInt), optional(schemas.BigInt))
```
we should have an EthAddress schema somewhere, i imagine this could panic later if this string is not well formed

schemas.EthAddress i think

Maddiaa0 reviewed (approved):
ill defer merge approval to palla as hes more familiar with this component + pending schema tightening.

this change is super well contained, nice work. Thanks for explaining the motivation extremely well in the description

spalladino commented on yarn-project/aztec-node/src/aztec-node/server.ts:
```diff
@@ -1115,6 +1115,10 @@ export class AztecNodeService implements AztecNode, AztecNodeAdmin, Traceable {
 
+  public getValidatorStats(validatorAddress: string, fromSlot?: bigint, toSlot?: bigint): Promise<SingleValidatorStats | undefined> {
+    return this.validatorsSentinel?.getValidatorStats(validatorAddress, fromSlot, toSlot) ?? Promise.resolve(undefined);
+  }
```
As @Maddiaa0 said, let's use `EthAddress` instead of plain strings for validator addresses

0x4r45h replied:
Done

spalladino commented on yarn-project/aztec-node/src/sentinel/sentinel.ts:
```diff
@@ -375,6 +376,35 @@ export class Sentinel extends (EventEmitter as new () => WatcherEmitter) impleme
+    }
+
+    const slotNow = this.epochCache.getEpochAndSlotNow().slot;
+    const effectiveFromSlot = fromSlot ?? (this.lastProcessedSlot ?? slotNow) - BigInt(this.store.getHistoryLength());
+    const effectiveToSlot = toSlot ?? this.lastProcessedSlot ?? slotNow;
```
We should check if the diff between from and to is greater than history length, and if so, throw. Otherwise, the caller will expect data for a given slot range, but get missing data in the response.

0x4r45h replied:
Added a check ;it throws an error on bad ranges.

spalladino commented on yarn-project/aztec-node/src/sentinel/sentinel.ts:
```diff
@@ -375,6 +376,35 @@ export class Sentinel extends (EventEmitter as new () => WatcherEmitter) impleme
+    const effectiveToSlot = toSlot ?? this.lastProcessedSlot ?? slotNow;
+
+    const validator = this.computeStatsForValidator(validatorAddress as `0x${string}`, history, effectiveFromSlot, effectiveToSlot);
+    const provenPerformance = await this.store.getProvenPerformance(ethAddress);
```
Note that proven performance is returned for the entire history, and not filtered by the slot range requested. Is this intentional? If so, let's rename to something that makes it clear, like `completeProvenPerformance` or `allProvenPerformance` (I don't really like those, so feel free to change if you come up with a better one!).

0x4r45h replied:
Nice catch, no it wasn't intentional. changed the name to `allTimeProvenPerformance`

spalladino commented on yarn-project/aztec-node/src/sentinel/sentinel.ts:
```diff
@@ -375,6 +376,35 @@ export class Sentinel extends (EventEmitter as new () => WatcherEmitter) impleme
   }
 
+  /** Computes stats for a single validator. */
+  public async getValidatorStats(
```
Mind adding a few tests for this function? Feel free to mock the result of `computeStatsForValidator` if it makes it easier.

0x4r45h replied:
Done

spalladino reviewed (commented):
Looks good, thanks for the contribution, and thanks for the detailed explanation! I added a few comments which should be straightforward to address.

0x4r45h committed:
Throw Error on bad slot range

use EthAddress Schema instead of string

better naming for provenPerformance field

add a few tests for getValidatorStats

0x4r45h committed:
Add more tests for getValidatorStats

spalladino reviewed (approved):
Looks solid, thanks for the contribution!!

spalladino commented:
Merged as #16446. Apologies for having to close this one @StoneMac65, we are having some issues with our CI for external contributors, and this was the fastest way to get your PR in. Rest assured you are flagged as author for your commits.

StoneMac65 commented:
> Merged as #16446. Apologies for having to close this one @StoneMac65, we are having some issues with our CI for external contributors, and this was the fastest way to get your PR in. Rest assured you are flagged as author for your commits.

no worries at all. we are glad it was helpful
